### ISSUE TITLE
LFX Workspace: Improving the performance of running rustpython

### ISSUE THREAD

**Issue by wenym1** [WasmEdge/WasmEdge#1293]
## Motivation

RustPython will be a practical use case on WasmEdge. Current RustPython has some known issues (such as #996 ) about soundness, robustness and performance when running on WasmEdge. This mentorship program will try to fix some issues about RustPython on WasmEdge and try to improve the performance under both AOT and non-AOT scenario for >7%.

## Milestone

- [x] Compile and build RustPython to `rustpython.wasm`
- [x] Build `wasmedge` and `wasmedgec` locally, use `wasmedgec` to transform `rustpython.wasm` to `rustpython-aot.wasm`, an AOT-optimized RustPython wasm file. Run `rustpython.wasm` and `rustpython-aot.wasm` on WasmEdge respectively.
- [x] Read the source code of WasmEdge and understand how it works.
- [ ] Fix #996 and figure out the reason why the program fails on exit.
- [x] Install `wasmtime` and `wasmer` and run RustPython with basic python code.
- [x] Develop several python codes for benchmark, including both computation-intensive and IO-intensive code
- [x] Develop and benchmark tools to easily run benchmark python codes on different wasm runtime and present the comparison in different dimension.
- [x] Use profiling tools to figure out the bottleneck of WasmEdge and try to improve the performance when RustPython is running on it. (Will add more details when we reach here)

**Comment by wenym1** [comment]
Till now we have compiled and built RustPython to rustpython.wasm, built wasmedge and wasmedgec locally and run rustpython with both AOT-optimized and raw wasm file.

The error on exit problem for AOT-optimized wasm file is reproduced locally. Currently we are working on this problem and try to comprehend how WasmEdge works by the way.

We have figured out that the runtime fails when trying to call a host function (haven't figure out which host function was called) on exit. The failure to call the host function trigger a Fault named `terminated` with error code 0x01 and in the end result to an error.

**Comment by wenym1** [comment]
Just analyze the reason why under AOT mode RustPython cannot exit cleanly in #1378

**Comment by wenym1** [comment]
Submitted an issue about build failure on MacOS with M1 chip. #1485

**Comment by wenym1** [comment]
Ran a simple rustpython benchmark with interpreter mode.  Following are some benchmark setting.
- Environment and steps to build the project is described in #1485.
- The wasm file of rustpython can be built according to the guild in [Github/RustPython](https://github.com/RustPython/RustPython).
- A generated wasm file of rustpython used in the benchmark is [rustpython.wasm](https://github.com/wenym1/WasmEdge/blob/master/rust-python-benchmark/rustpython.wasm).
- A compute-intensive python code is ran in the benchmark. [arithmetic_bench.py](https://github.com/wenym1/WasmEdge/blob/master/rust-python-benchmark/bench-src/arithmetic_bench.py)
To run the python code, simply ran `cat arithmetic_bench.py | time ./build/tools/wasmedge/wasmedge rustpython.wasm`.
On my Macbook Pro, the code in the master branch takes `119.07` seconds.
- I ran the `dtrace` profiling tools to trace the execution time in each function. 
```
cat arithmetic_bench.py | sudo dtrace -x 'ustackframes=100' -n 'profile-997 /pid == $target/ { @[ustack(100)] = count(); }' -o wasmedge.rust-python.stacks -c './build/tools/wasmedge/wasmedge rustpython.wasm'
```
The generated stack file is `wasmedge.rust-python.stacks`. We can use the tools in [Github/FlameGraph](https://github.com/brendangregg/FlameGraph) to generate the flamegraph to look for potential bottleneck.
The command is `./FlameGraph/stackcollapse.pl wasmedge.rust-python.stacks | ./FlameGraph/flamegraph.pl > wasmedge.rust-python.svg`. A generated flamegraph file is in ![`wasmedge.rust-python.svg`](https://github.com/wenym1/WasmEdge/blob/master/rust-python-benchmark/result/wasmedge.rust-python.svg).

**Comment by juntao** [comment]
Thank you. Did you run the AOT compiler `wasmedgec` on the `rustpython.wasm` file? Thanks.

**Comment by wenym1** [comment]
Besides, I use a hashmap to collect the count of each op code while running the benchmark.
The count result is in [op_code_count.txt](https://github.com/wenym1/WasmEdge/blob/master/rust-python-benchmark/result/op_code_count.txt).

The following op code are some most frequently used op code. The op code are sorted in ascending order. On the right of the op code are their portion to all op.
```
66 0.0113244697893
41 0.0119144019311
113 0.0127962290866
55 0.0131004104031
54 0.0254018148157
33 0.0285163419989
11 0.0362771799532
40 0.0384728288879
13 0.0445827694778
34 0.0533052006967
106 0.0695898665754
65 0.115945790243
32 0.203425308808
2 0.212691022498
```

**Comment by wenym1** [comment]
From the flamegraph and the op count statistics, I suppose that loading and storing memory and calling host function can possibly become a bottleneck.

For loading and storing memory, from the flamegraph, we can see that some execution time is spent on `std::memcpy`. I looked into the call stack and figured out that it is called in `loadValue` and `storeValue` in [memory.h](https://github.com/WasmEdge/WasmEdge/blob/master/include/runtime/instance/memory.h). When we look upper to the call stack, I find that most call to these two methods take 1, 2, 4, 8 or some other power of 2 as the `Length` argument, and such `Length` is the length argument passed to `std::memcpy`. However, `std::memcpy` is not efficient enough, as it only reads one byte at a time. Therefore, we can replace a simple `std::memcpy` with something like the following.
```
switch (Length) {
  case 1: *reinterpret_cast<uint8_t*>(&LoadVal) = *reinterpret_cast<uint8_t*>(&DataPtr[Offset]);break;
  case 2: *reinterpret_cast<uint16_t*>(&LoadVal) = *reinterpret_cast<uint16_t*>(&DataPtr[Offset]);break;
  case 4: *reinterpret_cast<uint32_t*>(&LoadVal) = *reinterpret_cast<uint32_t*>(&DataPtr[Offset]);break;
  case 8: *reinterpret_cast<uint64_t*>(&LoadVal) = *reinterpret_cast<uint64_t*>(&DataPtr[Offset]);break;
  default: std::memcpy(&LoadVal, &DataPtr[Offset], Length); break;
}
```

For calling host function, I noticed that in the flamegraph, a portion of execution time are spent on mutex lock and unlock operation. The call stack is 
```
Executor::runCallOp
Expect<FunctionInstance *> getFunc(uint32_t Idx) const noexcept
std::shared_lock Lock(Mutex)
```
However, it seems that function instances are only mutable in initialization, and while the runtime is running wasm file, the function instances are immutable. Therefore, while running wasm file, we don't have to acquire the lock. We can remove locking the mutex in `Expect<FunctionInstance *> getFunc(uint32_t Idx) const noexcept`.

Tentatively, I apply the two optimization mentioned above, and the time taken to finish the benchmark has dropped from `119.07` seconds to `111.57` seconds, which is about 6.7% improvement.
